index/segment/mem/simple_terms_dict.go

```diff
@@ -0,0 +1,198 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"regexp"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+)
+
+// simpleTermsDictionary uses two-level map to model a terms dictionary.
+// i.e. fieldName -> fieldValue -> postingsList
+type simpleTermsDictionary struct {
+	fieldNamesLock sync.RWMutex
+	fieldNames     map[string]*fieldValuesMap
+
+	opts Options
+}
+
+func newSimpleTermsDictionary(opts Options) termsDictionary {
+	return &simpleTermsDictionary{
+		fieldNames: make(map[string]*fieldValuesMap, opts.InitialCapacity()),
+		opts:       opts,
+	}
+}
+
+func (t *simpleTermsDictionary) Insert(field doc.Field, i segment.DocID) error {
+	fieldName := string(field.Name)
+	fieldValues := t.getOrAddFieldName(fieldName)
+	fieldValue := string(field.Value)
+	return fieldValues.addDocIDForValue(fieldValue, i)
+}
+
+func (t *simpleTermsDictionary) Fetch(
+	fieldName []byte,
+	fieldValueFilter []byte,
+	isRegexp bool,
+) (segment.PostingsList, error) {
+	// check if we know about the field name
+	t.fieldNamesLock.RLock()
+	fieldValues, ok := t.fieldNames[string(fieldName)]
+	t.fieldNamesLock.RUnlock()
+	if !ok {
+		// not an error to not have any matching values
+		return nil, nil
+	}
+
+	// get postingList(s) for the given value.
+	lists, err := fieldValues.fetchLists(fieldValueFilter, isRegexp)
+	if err != nil {
+		return nil, err
+	}
+
+	// union all the postingsList(s)
+	unionedList := t.opts.PostingsListPool().Get()
+	for _, ids := range lists {
+		unionedList.Union(ids)
+	}
+
+	return unionedList, nil
+}
+
+func (t *simpleTermsDictionary) getOrAddFieldName(fieldName string) *fieldValuesMap {
+	// cheap read lock to see if it already exists
+	t.fieldNamesLock.RLock()
+	fieldValues, ok := t.fieldNames[fieldName]
+	t.fieldNamesLock.RUnlock()
+	if ok {
+		return fieldValues
+	}
+
+	// acquire write lock and create
+	t.fieldNamesLock.Lock()
+	fieldValues, ok = t.fieldNames[fieldName]
+
+	// check if it's been created since we last acquired the lock
+	if ok {
+		t.fieldNamesLock.Unlock()
+		return fieldValues
+	}
+
+	fieldValues = newFieldValuesMap(t.opts)
+	t.fieldNames[fieldName] = fieldValues
+	t.fieldNamesLock.Unlock()
+	return fieldValues
+}
+
+type fieldValuesMap struct {
+	sync.RWMutex
+
+	opts Options
+
+	// fieldValue -> postingsList
+	values map[string]segment.PostingsList
+}
+
+func newFieldValuesMap(opts Options) *fieldValuesMap {
+	return &fieldValuesMap{
+		opts:   opts,
+		values: make(map[string]segment.PostingsList),
+	}
+}
+
+func (f *fieldValuesMap) addDocIDForValue(value string, i segment.DocID) error {
+	// try read lock to see if we already have a postingsList for the given value.
+	f.RLock()
+	pid, ok := f.values[value]
+	f.RUnlock()
+
+	// we have a postingsList, mark the docID and move on.
+	if ok {
+		pid.Insert(i)
+		return nil
+	}
+
+	// postingsList doesn't exist, time to acquire write lock
+	f.Lock()
```

Comment:
Seems cheap enough here to `defer ...Unlock()`.

Reply:
same as above, will change if you feel strongly.